### ISSUE TITLE
irrefutable `with` patterns in the scope of a `where` clause

### ISSUE THREAD

**Issue by jamesmckinna** [agda/agda#5883]
I previously posted this on the Agda Zulip, #**newcomers>`where` clauses governing irrefutable `with` clauses** , but no-one answered (well, Matthew did, but he wasn't familiar with nested `with`...), and the workaround, using a parenthesised `let`, seems unsatisfactory in general. Is this 'merely' a parsing issue, or something more subtle going on?

Suppose I have a function defined by nested with clauses
```
f x with g x
... | p with h p q
  where q = ...
... | r = t
```
where in the second `with` I supply some locally defined expression `q` governed by a `where` clause, which mentions names bound in the pattern `p`. So far so good, provided I remember to put the `where` clause in the right place, ie before the pattern introduced by the second, nested `with` clause.

But suppose now the patterns `p` and `r` above are irrefutable... and I try to write
```
f x with p <- g x
    with r <- h p q
        where q = ...
  = t
```
I seem to get a parse error in the above formulation; similarly, if I try instead to disambiguate scope by writing
```
f x with p <- g x
    with r <- (h p q where q = ...)
  = t
```
and finally if I put the `where` clause after the final `= t`
```
f x with p <- g x
    with r <- h p q
  = t
  where q = ...
```
then I get a scope error saying
"not in scope q when scope checking q"

I can solve the problem by using an explicit `let ...in ...` in place of the `where` clause, which seems to do the trick, but unless it is parenthesised,  the parser throws an "unexpected '=' in... " error.
```
f x with p <- g x
    with r <- (let q = ... in h p q)
  = t
```
Is this a bug? A feature? Is there a palatable solution (which localises the scope of the `where` clause to the irrefutable `with` clause which mentions the bound name) besides resorting to 'old-style' `with`-notation with ellipses, or the parenthesised `let`? I've grown rather fond of irrefutable (esp. nested) `with`, but find my problems with the above scenario a serious limitation on its use.

I can provide a (non-)working example (taken from some lecture notes I am writing for a Summer School; it's taken from an example result in operational semantics, ... with the above call pattern arising from successive recursive calls on an evaluation relation on successive subterms of the top-level pattern, threading some environment updates between them), but it's not (easy for me to extract) a minimal one, so I've resorted to generalities/hypotheticals in the above description. But as a scenario, I imagine it to be the kind of thing which might crop up all the time in mechanised metatheory proofs, and where a simple `let` binding might not be the most perspicuous form in which to present the proof...

**Comment by jamesmckinna** [comment]
UPDATED: the solution I arrived at at AIM XXXVI was simply to replace
```agda
f x with p <- g x
    with r <- (let q = ... in h p q)
  = t
```
with an additional (non-matching) irrefutable `with` to simulate the `let` binding
```agda
f x with p <- g x
    with q <- ...
    with r <- h p q
  = t
```
which solves my particular problem.

On that basis, I'd be happy to close the issue, but the more general problem of having a `where` block scope over a subexpression on which an irrefutable `with` matches, remains open.

**Comment by jamesmckinna** [comment]
I raised this at AIMXXXVI, and I think there was some discussion of how allowing (possibly pattern-matching) `let` on the LHS would solve this: what's the current status of such an enhancement? A merged implementation for v2.6.4?

**Comment by jamesmckinna** [comment]
Suggest closing now: I have a good-enough solution, and better might follow one day.
